server/src/Lib.hs

```diff
@@ -0,0 +1,200 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Lib (app) where
+
+import Control.Monad
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Except (ExceptT (..), runExceptT, withExceptT)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import Data.Coerce (coerce)
+import Data.Either (isLeft)
+import Data.Maybe (fromMaybe)
+import Database.SQLite.Simple
+import Database.SQLite.Simple.FromField (FromField (..))
+import Database.SQLite.Simple.ToField (ToField (..))
+import GHC.IO.Exception (ExitCode (ExitSuccess))
+import Network.Wai.Handler.Warp (run)
+import Servant
+import qualified System.Directory as Dir
+import System.FilePath ((</>))
+import qualified System.Process as Proc
+
+-- | Top-level API. Examples:
+--   localhost:8081/ps
+--   localhost:8081/procps/ps?sys=aarch64-linux
+--  TODO API documentation
+type API =
+  Get '[JSON] NoContent
+    :<|> Capture "binary-name" BinaryName :> QueryParam "sys" Platform :> Get '[OctetStream] ByteString
+    :<|> Capture "package-name" PackageName :> Capture "binary-name" BinaryName :> QueryParam "sys" Platform :> Get '[OctetStream] ByteString
+
+newtype BinaryName = BinaryName {unBinaryName :: String}
+  deriving newtype (ToField, FromHttpApiData)
+
+newtype PackageName = PackageName {unPackageName :: String}
+  deriving newtype (Eq, Ord, FromField, FromHttpApiData, ToField)
+
+data Platform
+  = X86_64_Linux
+  | AARCH_64_Linux
+
+instance Show Platform where
+  show X86_64_Linux = "x86_64-linux"
+  show AARCH_64_Linux = "aarch64-linux"
+
+instance FromHttpApiData Platform where
+  parseQueryParam "aarch64" = pure AARCH_64_Linux
+  parseQueryParam "x86_64-linux" = pure X86_64_Linux
+  parseQueryParam sys = Left $ "Unknown platform: " <> sys
+
+-- | A Triplet uniquely identifies a binary in nixpkgs.
+-- A Triplet is not a proof that that binary exists and is buildable, though.
+data Triplet = Triplet
+  { _tripBinary :: BinaryName,
+    _tripPackage :: PackageName,
+    _tripPlatform :: Platform
+  }
```

Comment:
I was thinking about this a little, and one thing I realized is that the same binary can sometimes be exposed through multiple packages (like if a package has an alias).

For instance, the top-level attributes `git-annex` and `haskellPackages.git-annex` both give you the same derivation (with the same binaries), even though their `PackageName` would be different.

I guess it would be nice if we somehow took this into account, But honestly I don't have a good idea here.  I think `Triplet` here is good.  Maybe at some point in the future we'd figure out some way to improve it.

Reply:
At some point we probably don't want to invoke a nix build call for every request, and instead link a triplet directly to a path in either the nix store or some other caching mechanism. If that path is derivation/content-addressed, we'd get this deduplication for free, and we'd also be able to query the database for what paths have multiple associated triplets.

Reply:
Oh nice, that's a good idea.